{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "54213091",
   "metadata": {},
   "source": [
    "# Change Gene and Contig Names in pyriformis GFF \n",
    "\n",
    "\n",
    "1. Open and parse GFF file to pull out information from each column\n",
    "\n",
    "2. Rename the contigs in column 1 \n",
    "\n",
    "3. Rename the genes in column 9 (attributes column, ID = \"StePyr_1\" etc)\n",
    "\n",
    "4. Rename the ID and Parent in column 9 for transcript, CDS, intron\n",
    "\n",
    "5. Write file\n",
    "\n",
    "       \n",
    "\n",
    " \n",
    " \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23ad154e",
   "metadata": {},
   "source": [
    "## 1. GFF Parser\n",
    "courtesy of Colleen Hannon and Mike Eisen, edited for the format of my GFF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "32b77771",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use class to define the gff object\n",
    "# It seperates each column into a particular attribute of the gff class\n",
    "\n",
    "class gff:\n",
    "    \n",
    "    def __init__(self):\n",
    "        \n",
    "        self.seq = \"\"\n",
    "        self.source = \"\"\n",
    "        self.feature = \"\"\n",
    "        self.start = \"\"\n",
    "        self.end = \"\"\n",
    "        self.score = \"\"\n",
    "        self.strand = \"\"\n",
    "        self.frame = \"\"\n",
    "        self.attribute = {}\n",
    "        \n",
    "    def __iter__(self): \n",
    "        return iter([self.seq, self.source, self.feature, self.start, self.end, self.score, self.strand, self.frame, self.attribute])\n",
    "\n",
    "## This is a function that reads in each row of the gff, and pases them into the proper \n",
    "## Attribute class\n",
    "\n",
    "    @classmethod\n",
    "    def from_fields(cls,fields):\n",
    "        gff = cls()\n",
    "        if len(fields) == 9:\n",
    "            gff.seq = fields[0]\n",
    "            gff.source = fields[1]\n",
    "            gff.feature = fields[2]\n",
    "            gff.start = int(fields[3])\n",
    "            gff.end = int(fields[4])\n",
    "            if fields[5] != \".\":\n",
    "                gff.score = float(fields[5])\n",
    "            gff.strand = fields[6]\n",
    "            gff.frame = fields[7]\n",
    "            \n",
    "            atts = fields[8].split(';')\n",
    "            \n",
    "            while(\"\" in atts):\n",
    "                atts.remove(\"\")\n",
    "        \n",
    "            for att in atts:\n",
    "                (k,v) = att.split('=')\n",
    "                gff.attribute[k] = v\n",
    "        return gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a9adfdeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_file = \"/Volumes/albright_postdoc/pyriformis/stentor_pyriformis.20210302_final.0.2.final.proper.gff\"\n",
    "\n",
    "gffs = {}\n",
    "\n",
    "## Only store the ones in these category.\n",
    "\n",
    "features = [\n",
    "              'CDS',\n",
    "              'enhancer',\n",
    "              'exon',\n",
    "              'five_prime_UTR',\n",
    "              'gene',\n",
    "              'golden_path_region',\n",
    "              'insulator',\n",
    "              'intergenic',\n",
    "              'intron',\n",
    "              'mRNA',\n",
    "              'mature_peptide',\n",
    "              'miRNA',\n",
    "              'ncRNA',\n",
    "              'none',\n",
    "              'polyA_site',\n",
    "              'pre_miRNA',\n",
    "              'protein',\n",
    "              'pseudogene',\n",
    "              'rRNA',\n",
    "              'regulatory_region',\n",
    "              'repeat_region',\n",
    "              'silencer',\n",
    "              'snRNA',\n",
    "              'snoRNA',\n",
    "              'tRNA',\n",
    "              'tandem_repeat',\n",
    "              'three_prime_UTR',\n",
    "              'transposable_element',\n",
    "              'transcript'\n",
    "              ]\n",
    "\n",
    "for feature in features:\n",
    "    gffs[feature] = []\n",
    "\n",
    "gffo = open(gff_file,\"r\", newline = '')             \n",
    "\n",
    "for line in gffo:\n",
    "    if line[0:1] != \"#\":\n",
    "        line = line.strip('\\n')\n",
    "        line = line.split(\"\\t\")\n",
    "        if len(line) == 9: \n",
    "            if line[2] in features: ## makes sure the row is of a feature we are interested in  \n",
    "                gffs[line[2]].append(gff.from_fields(line))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0322294",
   "metadata": {},
   "source": [
    "## 2. Rename contigs\n",
    "\n",
    "Contig names are a ugly mess of a string, make it nice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d7473109",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a list of unique contig names\n",
    "contig_list = []\n",
    "\n",
    "for line in gffs['gene']: \n",
    "     if line.seq not in contig_list:\n",
    "            contig_list.append(line.seq)\n",
    "            \n",
    "# generate a list of new contig names\n",
    "base = 'StePyr_contig'\n",
    "new_contig_list = [\"{}_{}\".format(base, (i+1)) for i in range(0, len(contig_list))]\n",
    "\n",
    "# generate a dictionary with unique contig names mapped to new name\n",
    "contig_dict = dict(map(lambda i,j : (i,j) , contig_list,new_contig_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9bbb8aa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for every feature present in GFF file, replace contig name (first column of GFF file)\n",
    "\n",
    "for feature in features: \n",
    "    for line in gffs[feature]: \n",
    "        line.seq = (line.seq if line.seq not in contig_dict else contig_dict[line.seq])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "583c1610",
   "metadata": {},
   "source": [
    "## 3. Rename genes \n",
    "\n",
    "\n",
    "My GFF file had the contig as the gene ID, meaning I had to fix non-unique names\n",
    "\n",
    "Some of this is from Colleen and Mike, some by me"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b0eba7d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "class gene:\n",
    "    \n",
    "    def __init__(self):\n",
    "        \n",
    "        self.id = \"\"\n",
    "        self.children = []\n",
    "    \n",
    "    @classmethod\n",
    "    def from_atts(cls,atts):\n",
    "        gene = cls()\n",
    "        if 'ID' in atts:\n",
    "            gene.id = atts['ID']\n",
    "        return gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "10093108",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "new_genes_list = []\n",
    "\n",
    "for gene_gff in gffs['gene']:\n",
    "\n",
    "    atty = gene_gff.attribute\n",
    "    \n",
    "    i += 1\n",
    "    \n",
    "    atty['ID'] = 'StePyr_' + str(i)\n",
    "    \n",
    "    new_genes_list.append(atty['ID'])\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d15d741c",
   "metadata": {},
   "source": [
    "## 4. Rename transcripts, CDS, introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f63e8274",
   "metadata": {},
   "outputs": [],
   "source": [
    "class transcript: \n",
    "    \n",
    "    def __init__(self): \n",
    "        \n",
    "        self.id = \"\"\n",
    "        self.parent = \"\"\n",
    "        self.children = []\n",
    "        \n",
    "    @classmethod\n",
    "    def from_atts(cls,atts):\n",
    "        transcript = cls()\n",
    "        if 'ID' in atts: \n",
    "            transcript.id = atts['ID']\n",
    "        if 'Parent' in atts: \n",
    "            transcript.parent = atts['Parent']\n",
    "        return transcript\n",
    "    \n",
    "    \n",
    "class CDS: \n",
    "    \n",
    "    def __init__(self):\n",
    "        \n",
    "        self.id = \"\"\n",
    "        self.parent = \"\"\n",
    "        \n",
    "    @classmethod\n",
    "    def from_atts(cls,atts):\n",
    "        CDS = cls()\n",
    "        if 'ID' in atts: \n",
    "            CDS.id = atts['ID']\n",
    "        if 'Parent' in atts: \n",
    "            CDS.parent = atts['Parent']\n",
    "        return CDS \n",
    "    \n",
    "    \n",
    "class intron: \n",
    "    \n",
    "    def __init__(self):\n",
    "        \n",
    "        #self.id = \"\"\n",
    "        self.parent = \"\"\n",
    "        \n",
    "    @classmethod\n",
    "    def from_atts(cls,atts):\n",
    "        intron = cls()\n",
    "        #if 'ID' in atts:\n",
    "            #intron.id = atts['ID'][0]\n",
    "        if 'Parent' in atts: \n",
    "            intron.parent = atts['Parent']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdda9f38",
   "metadata": {},
   "source": [
    "### Transcripts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6022a6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "old_transcript_list = []\n",
    "\n",
    "for transcript_gff in gffs['transcript']: \n",
    "    atty = transcript_gff.attribute\n",
    "    old_transcript_list.append(atty['ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "50057bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_transcript_list = []\n",
    "\n",
    "for i in new_genes_list: \n",
    "    new_transcript_list.append(i + '.t1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c553dd23",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_transcript_dict = dict(zip(old_transcript_list, new_transcript_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2d7055f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for transcript_gff in gffs['transcript']:\n",
    "   \n",
    "    atty = transcript_gff.attribute\n",
    "    \n",
    "    for k,v in new_transcript_dict.items(): \n",
    "    \n",
    "        atty['ID'] = atty['ID'].replace(k,v)\n",
    " \n",
    "    atty['Parent'] = atty['ID'][:-3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57910221",
   "metadata": {},
   "source": [
    "### CDSs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1e219a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "for CDS_gff in gffs['CDS']:\n",
    "   \n",
    "    atty = CDS_gff.attribute\n",
    "    \n",
    "    for k,v in new_transcript_dict.items(): \n",
    "    \n",
    "        atty['ID'] = atty['ID'].replace(k,v)\n",
    " \n",
    "    atty['Parent'] = atty['ID'][:-4]   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d480fedb",
   "metadata": {},
   "source": [
    "### Introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9a81b360",
   "metadata": {},
   "outputs": [],
   "source": [
    "for intron_gff in gffs['intron']:\n",
    "    \n",
    "    atty = intron_gff.attribute\n",
    "    \n",
    "    for k,v in new_transcript_dict.items(): \n",
    "    \n",
    "        #atty['ID'] = atty['ID'].replace(k,v)\n",
    " \n",
    "        atty['Parent'] = atty['Parent'].replace(k,v)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b7f86a9",
   "metadata": {},
   "source": [
    "## 5. Write file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c8aa6926",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "db328ca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"draft_gff.csv\", \"w\") as stream:\n",
    "    writer = csv.writer(stream, delimiter = '\\t')\n",
    "    writer.writerows(gffs['CDS'] + gffs['gene'] + gffs['transcript'] + gffs['intron'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ec23ae4",
   "metadata": {},
   "source": [
    "## 6. Import file, fix it, write again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e7686f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    " \n",
    "# making data frame from csv file\n",
    "data = pd.read_csv(\"draft_gff.csv\", sep = '\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9a81b093",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.sort_values([0,3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7ca8a3e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# maybe there's a neater way to do this, but whatever it fixes the format!\n",
    "\n",
    "data[8] = data[8].str.replace('{', '')\n",
    "data[8] = data[8].str.replace('}', '')\n",
    "data[8] = data[8].str.replace(',', ';')\n",
    "data[8] = data[8].str.replace(':', '=')\n",
    "data[8] = data[8].str.replace(' ', '')\n",
    "data[8] = data[8].str.replace('\\'', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b3e1d775",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>StePyr_contig_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9537</td>\n",
       "      <td>19184</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=StePyr_1.t1.cds;Parent=StePyr_1.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19199</th>\n",
       "      <td>StePyr_contig_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>9537</td>\n",
       "      <td>19184</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=StePyr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34830</th>\n",
       "      <td>StePyr_contig_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>9537</td>\n",
       "      <td>19184</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=StePyr_1.t1;Parent=StePyr_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>StePyr_contig_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>19876</td>\n",
       "      <td>33888</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=StePyr_2.t1.cds;Parent=StePyr_2.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19200</th>\n",
       "      <td>StePyr_contig_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>19876</td>\n",
       "      <td>33888</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=StePyr_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12486</th>\n",
       "      <td>StePyr_contig_99</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>226886</td>\n",
       "      <td>226954</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=StePyr_10134.t1.cds;Parent=StePyr_10134.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29332</th>\n",
       "      <td>StePyr_contig_99</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>226886</td>\n",
       "      <td>227650</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=StePyr_10134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44963</th>\n",
       "      <td>StePyr_contig_99</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>226886</td>\n",
       "      <td>227650</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=StePyr_10134.t1;Parent=StePyr_10134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52814</th>\n",
       "      <td>StePyr_contig_99</td>\n",
       "      <td>INTRONARRATOR</td>\n",
       "      <td>intron</td>\n",
       "      <td>226955</td>\n",
       "      <td>226969</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=StePyr_10134.t1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12487</th>\n",
       "      <td>StePyr_contig_99</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>CDS</td>\n",
       "      <td>226970</td>\n",
       "      <td>227650</td>\n",
       "      <td>1.0</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=StePyr_10134.t1.cds;Parent=StePyr_10134.t1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54029 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      0              1           2       3       4    5  6  7  \\\n",
       "0       StePyr_contig_1       AUGUSTUS         CDS    9537   19184  1.0  +  0   \n",
       "19199   StePyr_contig_1       AUGUSTUS        gene    9537   19184  1.0  +  .   \n",
       "34830   StePyr_contig_1       AUGUSTUS  transcript    9537   19184  1.0  +  .   \n",
       "1       StePyr_contig_1       AUGUSTUS         CDS   19876   33888  1.0  +  0   \n",
       "19200   StePyr_contig_1       AUGUSTUS        gene   19876   33888  1.0  +  .   \n",
       "...                 ...            ...         ...     ...     ...  ... .. ..   \n",
       "12486  StePyr_contig_99       AUGUSTUS         CDS  226886  226954  1.0  +  0   \n",
       "29332  StePyr_contig_99       AUGUSTUS        gene  226886  227650  1.0  +  .   \n",
       "44963  StePyr_contig_99       AUGUSTUS  transcript  226886  227650  1.0  +  .   \n",
       "52814  StePyr_contig_99  INTRONARRATOR      intron  226955  226969  1.0  +  .   \n",
       "12487  StePyr_contig_99       AUGUSTUS         CDS  226970  227650  1.0  +  0   \n",
       "\n",
       "                                                   8  \n",
       "0              ID=StePyr_1.t1.cds;Parent=StePyr_1.t1  \n",
       "19199                                    ID=StePyr_1  \n",
       "34830                 ID=StePyr_1.t1;Parent=StePyr_1  \n",
       "1              ID=StePyr_2.t1.cds;Parent=StePyr_2.t1  \n",
       "19200                                    ID=StePyr_2  \n",
       "...                                              ...  \n",
       "12486  ID=StePyr_10134.t1.cds;Parent=StePyr_10134.t1  \n",
       "29332                                ID=StePyr_10134  \n",
       "44963         ID=StePyr_10134.t1;Parent=StePyr_10134  \n",
       "52814                         Parent=StePyr_10134.t1  \n",
       "12487  ID=StePyr_10134.t1.cds;Parent=StePyr_10134.t1  \n",
       "\n",
       "[54029 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0ad98ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv('/Volumes/albright_postdoc/pyriformis/stentor_pyriformis.20210302_final.0.2.final.proper.nameupdate.gff', sep = '\\t', index = False, header = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c484e8fa",
   "metadata": {},
   "source": [
    "# Final Check: \n",
    "\n",
    "Update contig names in the genome fasta to match these, compare to old names in IGV. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6624a0f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
